shell/platform/android/io/flutter/embedding/engine/android/AndroidTouchProcessor.java

```diff
@@ -57,27 +79,43 @@ public boolean onTouchEvent(MotionEvent event) {
 
     // Prepare a data packet of the appropriate size and order.
     ByteBuffer packet = ByteBuffer.allocateDirect(
-        pointerCount * POINTER_DATA_FIELD_COUNT * BYTE_PER_FIELD
+        pointerCount * POINTER_DATA_FIELD_COUNT * BYTES_PER_FIELD
     );
     packet.order(ByteOrder.LITTLE_ENDIAN);
 
     int maskedAction = event.getActionMasked();
-    // ACTION_UP, ACTION_POINTER_UP, ACTION_DOWN, and ACTION_POINTER_DOWN
-    // only apply to a single pointer, other events apply to all pointers.
-    if (maskedAction == MotionEvent.ACTION_UP || maskedAction == MotionEvent.ACTION_POINTER_UP
-        || maskedAction == MotionEvent.ACTION_DOWN || maskedAction == MotionEvent.ACTION_POINTER_DOWN) {
-      addPointerForIndex(event, event.getActionIndex(), packet);
+    int pointerChange = getPointerChangeForAction(event.getActionMasked());
+    if (maskedAction == MotionEvent.ACTION_DOWN || maskedAction == MotionEvent.ACTION_POINTER_DOWN) {
+      // ACTION_DOWN and ACTION_POINTER_DOWN always apply to a single pointer only.
+      addPointerForIndex(event, event.getActionIndex(), pointerChange, 0, packet);
+    } else if (maskedAction == MotionEvent.ACTION_UP || maskedAction == MotionEvent.ACTION_POINTER_UP) {
+      // ACTION_UP and ACTION_POINTER_UP may contain position updates for other pointers.
+      // We are converting these updates to move events here in order to preserve this data.
+      // We also mark these events with a flag in order to help the framework reassemble
+      // the original Android event later, should it need to forward it to a PlatformView.
+      for (int p = 0; p < pointerCount; p++) {
+        if (p != event.getActionIndex()) {
+          if (event.getToolType(p) == MotionEvent.TOOL_TYPE_FINGER) {
+            addPointerForIndex(event, p, PointerChange.MOVE, POINTER_DATA_FLAG_BATCHED, packet);
+          }
+        }
+      }
+      // It's important that we're sending the UP event last. This allows PlatformView
+      // to correctly batch everything back into the original Android event if needed.
+      addPointerForIndex(event, event.getActionIndex(), pointerChange, 0, packet);
     } else {
       // ACTION_MOVE may not actually mean all pointers have moved
       // but it's the responsibility of a later part of the system to
       // ignore 0-deltas if desired.
       for (int p = 0; p < pointerCount; p++) {
-        addPointerForIndex(event, p, packet);
+        addPointerForIndex(event, p, pointerChange, 0, packet);
       }
     }
 
     // Verify that the packet is the expected size.
-    assert packet.position() % (POINTER_DATA_FIELD_COUNT * BYTE_PER_FIELD) == 0;
+    if (packet.position() % (POINTER_DATA_FIELD_COUNT * BYTES_PER_FIELD) != 0) {
```

Comment:
I'd keep this as an assert with no runtime cost.  Consuming the wrong number of bytes per packet should never happen by design.

Reply:
Ok.  I changed it to this because someone changed it from an assert to an if-statement in the FlutterView code.  But I'll change it back.